components/region_cache_memory_engine/src/background.rs

```diff
@@ -1027,6 +1071,84 @@ impl RunnableWithTimer for BackgroundRunner {
     }
 }
 
+pub struct DeleteRangeRunner {
+    engine: Arc<RwLock<RangeCacheMemoryEngineCore>>,
+    // It is possible that when `DeleteRangeRunner` begins to delete a range, the range is being
+    // written by apply threads. In that case, we have to delay the delete range task to avoid race
+    // condition between them. Periodically, these delayed ranges will be checked to see if it is
+    // ready to be deleted.
+    delay_ranges: Vec<CacheRange>,
+}
+
+impl DeleteRangeRunner {
+    fn new(engine: Arc<RwLock<RangeCacheMemoryEngineCore>>) -> Self {
+        Self {
+            engine,
+            delay_ranges: vec![],
+        }
+    }
+
+    fn delete_ranges(&mut self, ranges: &[CacheRange]) {
+        let skiplist_engine = self.engine.read().engine();
+        for r in ranges {
+            println!("delete range {:?}", r);
+            skiplist_engine.delete_range(r);
+        }
+        self.engine
+            .write()
+            .mut_range_manager()
+            .on_delete_ranges(ranges);
+        #[cfg(test)]
+        flush_epoch();
+    }
+}
+
+impl Runnable for DeleteRangeRunner {
+    type Task = BackgroundTask;
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            BackgroundTask::DeleteRange(ranges) => {
+                let (mut ranges_to_delay, ranges_to_delete) = {
+                    let core = self.engine.read();
+                    let mut ranges_to_delay = vec![];
+                    let mut ranges_to_delete = vec![];
+                    for r in ranges {
+                        // If the range is overlapped with ranges in `ranges_being_written`, the
+                        // range has to be delayed to delete. See comment on `delay_ranges`.
+                        if core
+                            .range_manager
+                            .range_overlapped_with_ranges_being_written(&r)
+                        {
+                            ranges_to_delay.push(r);
+                        } else {
+                            ranges_to_delete.push(r);
+                        }
+                    }
+                    (ranges_to_delay, ranges_to_delete)
+                };
+                self.delay_ranges.append(&mut ranges_to_delay);
+                self.delete_ranges(&ranges_to_delete);
```

Comment:
Can it prevent ABA problems? E.g., delete_range `[a, c)` is somehow delayed, and `[a, c)` is reloaded.

Reply:
After [a, c) is evicted and before it completes the deletion, it is in the ranges_being_deleted which prevents any range overlapped with it being loaded.